system/jlib/jptree.cpp

```diff
@@ -531,6 +541,340 @@ const char *queryHead(const char *xpath, StringBuffer &head)
 
 ///////////////////
 
+static constexpr unsigned defaultSiblingMapThreshold = 100;
+static unsigned siblingMapThreshold = (unsigned)-1; // off until configuration default it on.
+
+void setPTreeMappingThreshold(unsigned threshold)
+{
+    /*
+    * NB: setPTreeMappingThreshold() will automatically be called via loadConfiguration
+    * Redefining this limit, will not effect existing maps, and should generally only be called once during startup.
+    */
+    if (0 == threshold)
+        threshold = (unsigned)-1;
+    siblingMapThreshold = threshold;
+}
+
+class CQualifierMap : public std::unordered_multimap<std::string, IPropertyTree *>
+{
+    typedef std::unordered_multimap<std::string, IPropertyTree *> PARENT;
+    std::unordered_set<std::string> lhsProps;
+    bool hasPtrMap = false;
+public:
+    CriticalSection crit;
+
+    void addLhs(const char *lhs)
+    {
+        lhsProps.insert(lhs);
+    }
+    bool handlesLhs(const char *lhs) const
+    {
+        auto it = lhsProps.find(lhs);
+        return it != lhsProps.end();
+    }
+    void setHasPtrMap()
+    {
+        hasPtrMap = true;
+    }
+    bool handlesPtrs() const { return hasPtrMap; }
+    std::unordered_set<std::string> &getLhsProps()
+    {
+        return lhsProps;
+    }
+    void insertEntry(IPropertyTree *tree, const char *lhs, const char *v)
+    {
+        // NB: more efficient to use std::string, make_pair and emplace
+        std::string key = lhs + std::string("=") + std::string(v);
+        PARENT::emplace(std::make_pair(key, tree));
+    }
+    bool removeEntry(IPropertyTree *tree, const char *lhs, const char *v)
+    {
+        VStringBuffer key("%s=%s", lhs, v);
+        auto range = PARENT::equal_range(key.str());
+        if (range.first == range.second)
+            return false;
+
+        auto it = range.first;
+        while (true)
+        {
+            if (it->second == tree)
+            {
+                it = PARENT::erase(it);
+                return true;
+            }
+            ++it;
+            if (it == range.second)
+                break;
+        }
+        throwUnexpected();
+    }
+    void removeEntryIfMapped(IPropertyTree *tree, const char *lhs, const char *v)
+    {
+        if (lhsProps.find(lhs) != lhsProps.end())
+            removeEntry(tree, lhs, v);
+    }
+    void insertEntryIfMapped(IPropertyTree *tree, const char *lhs, const char *v)
+    {
+        if (lhsProps.find(lhs) != lhsProps.end())
+            insertEntry(tree, lhs, v);
+    }
+    void replaceEntryIfMapped(IPropertyTree *tree, const char *lhs, const char *oldv, const char *newv)
+    {
+        if (lhsProps.find(lhs) != lhsProps.end())
+        {
+            removeEntry(tree, lhs, oldv);
+            insertEntry(tree, lhs, newv);
+        }
+    }
+};
+
+bool parseEqualityQualifier(const char *&xxpath, unsigned &lhsLen, const char *&rhsBegin, unsigned &rhsLen)
```

Comment:
This looks like it should be commoned up with an existing function, but I haven't checked at all.

Reply:
it is similar, and there is duplication, but it is specialized enough, that I don't wanted to common up with checkPattern